src/eventHandlers/ResourceBookingEventHandler.js

```diff
@@ -124,6 +126,167 @@ async function assignJob (payload) {
+    })
+  }
+  if (workPeriodsToAdd.length > 0) {
+    await _createWorkPeriods(workPeriodsToAdd, payload.value.id)
+    logger.debug({
+      component: 'ResourceBookingEventHandler',
+      context: 'updateWorkPeriods',
+      message: `New WorkPeriods created for resource booking with id: ${payload.value.id}`
+    })
+  }
+  if (workPeriodsToUpdate.length > 0) {
+    await _updateWorkPeriods(workPeriodsToUpdate)
+    logger.debug({
+      component: 'ResourceBookingEventHandler',
+      context: 'updateWorkPeriods',
+      message: `WorkPeriods updated for resource booking with id: ${payload.value.id}`
+    })
+  }
+}
+
+/**
+ * When a ResourceBooking is deleted, workPeriods related to
+ * that ResourceBooking should also be deleted
+ * @param {object} payload the event payload
+ * @returns {undefined}
+ */
+async function deleteWorkPeriods (payload) {
+  // find related workPeriods to delete
+  const workPeriods = await WorkPeriod.findAll({
+    where: {
+      resourceBookingId: payload.value.id
+    },
+    raw: true
+  })
+  if (workPeriods.length === 0) {
+    logger.debug({
+      component: 'ResourceBookingEventHandler',
+      context: 'deleteWorkPeriods',
+      message: `id: ${payload.value.id} resource booking has no workPeriods - ignored`
+    })
+    return
+  }
+  await _deleteWorkPeriods(workPeriods)
+  logger.debug({
+    component: 'ResourceBookingEventHandler',
+    context: 'deleteWorkPeriods',
+    message: `WorkPeriods deleted for resource booking with id: ${payload.value.id}`
+  })
+}
+
+/**
+ * Calls WorkPeriodService to create workPeriods
+ * @param {Array<{startDate:Date,
+ * endDate:Date, daysWorked:number}>} periods work period data
+ * @param {string} resourceBookingId resourceBookingId of work period
+ * @returns {undefined}
+ */
+async function _createWorkPeriods (periods, resourceBookingId) {
+  Promise.all(_.forEach(periods, async period => await WorkPeriodService.createWorkPeriod(helper.getAuditM2Muser(),
+    {
+      resourceBookingId: resourceBookingId,
+      startDate: period.startDate,
+      endDate: period.endDate,
+      daysWorked: period.daysWorked,
+      paymentStatus: 'pending'
+    })))
+}
+
+/**
+ * Calls WorkPeriodService to update workPeriods
+ * @param {Array<{daysWorked:number}>} periods work period data
+ * @returns {undefined}
+ */
+async function _updateWorkPeriods (periods) {
+  Promise.all(_.forEach(periods, async period => await WorkPeriodService.partiallyUpdateWorkPeriod(helper.getAuditM2Muser(),
+    period.id,
+    {
+      daysWorked: period.daysWorked
+    })))
+}
+
+/**
+ * Calls WorkPeriodService to delete workPeriods
+ * @param {Array<{id:string}>} workPeriods work period objects
+ * @returns {undefined}
+ */
+async function _deleteWorkPeriods (workPeriods) {
+  Promise.all(_.forEach(workPeriods,
+    async workPeriod => await WorkPeriodService.deleteWorkPeriod(helper.getAuditM2Muser(), workPeriod.id)))
+}
```

Comment:
@eisbilir It seems for me that `Promise.all(_.forEach(workPeriods , async () => {...}))` would not work as expected, see https://stackoverflow.com/questions/37576685/using-async-await-with-a-foreach-loop

- `_deleteWorkPeriods` would **not** wait for all the async operation to get completed
- if some error happens during any async operation it would **not** be caught anywhere - we would not know about it.

Similar to this in other methods `_updateWorkPeriods` and `_createWorkPeriods`.

Reply:
@maxceem Yes it won't wait to complete and return immediately after firing all delete calls but errors will be caught if happens. It was my intention to not wait, because it is the very last step of all three operations (`_createWorkPeriods`, `_updateWorkPeriods`, `_deleteWorkPeriods`) and we don't stop processing if one of the calls raise error. (At least for now.)
But we can make it wait for all the async operation to get completed, so logger debug messages will appear in order.

Reply:
@eisbilir let's wait for the completion of these operations. As we do this in all other places, and such behaviour might be unexpected here.

In fact, our EventHandler here was designed in such a way, that it waits for all the handlers. So we decide that we need to not wait for some handlers to complete we have to implement in some very clear way.